lib/naver/searchad/api/core/http_command.rb

```diff
@@ -0,0 +1,139 @@
+            success(response, &block)
+          rescue => e
+            logger.debug("Error - #{e.inspect}")
+            error(e, &block)
+          end
+
+          def prepare!
+            normalize_unicode = true
+            if options
+              @header.merge!(options.header) if options.header
+              normalize_unicode = options.normalize_unicode
+            end
+
+            if url.is_a?(Addressable::Template)
+              @url = url.expand(params, nil, normalize_unicode)
+              @url.query_values = query.merge(url.query_values || {})
+            end
+
+            @body = '' unless body
+          end
+
+          def process_response(status, header, body)
+            check_status(status, header, body)
+            decode_response_body(header['Content-Type'].first, body)
+          end
+
+          def check_status(status, header = nil, body = nil, message = nil)
+            case status
+            when 200...300
+              nil
+            when 301, 302, 303, 307
+              raise Naver::Searchad::Api::RedirectError.new(
+                "Redirect to #{header['Location']}", status_code: status, header: header, body: body)
+            when 401
+              raise Naver::Searchad::Api::AuthorizationError.new(
+                'Unauthorized', status_code: status, header: header, body: body)
+            when 429
+              raise Naver::Searchad::Api::RateLimitError.new(
+                'Rate limit exceeded', status_code: status, header: header, body: body)
+            when 400, 402...500
+              raise Naver::Searchad::Api::RequestError.new(
+                'Invalid request', status_code: status, header: header, body: body)
+            when 500...600
+              raise Naver::Searchad::Api::ServerError.new(
+                'Server error', status_code: status, header: header, body: body)
+            else
+              logger.warn("Encountered unexpected status code #{status}")
+              raise Naver::Searchad::Api::UnknownError.new(
+                'Unknown error', status_code: status, header: header, body: body)
+            end
+          end
+
+          def decode_response_body(content_type, body)
+            body
+          end
+
+          private
+
+          def apply_request_options(req_header)
+            options.authorization.apply(req_header) if options.authorization.respond_to?(:apply)
+            req_header.merge!(header)
+          end
+
+          def success(result, &block)
+            block.call(result, nil) if block_given?
+            result
+          end
+
+          def error(err, &block)
+            if err.is_a?(HTTPClient::BadResponseError)
+              begin
+                res = err.res
+                check_status(res.status.to_i, res.header, res.body)
+              rescue Naver::Searchad::Api::Error => e
+                err = e
+              end
+            elsif err.is_a?(HTTPClient::TimeoutError) || err.is_a?(SocketError)
```

Comment:
Is `SocketError` always a timeout error? (I'm thinking about other errors like `SocketError: getaddrinfo: Name or service not known`)